python/lsst/daf/butler_migrate/script/rewrite_sqlite_registry.py

```diff
@@ -0,0 +1,255 @@
+# This file is part of daf_butler.
+#
+# Developed for the LSST Data Management System.
+# This product includes software developed by the LSST Project
+# (http://www.lsst.org).
+# See the COPYRIGHT file at the top-level directory of this distribution
+# for details of code ownership.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+__all__ = ("rewrite_sqlite_registry",)
+
+import io
+import os
+import logging
+from collections import defaultdict
+from typing import Dict
+import tempfile
+
+from lsst.daf.butler import (Config,
+                             Butler,
+                             DatasetAssociation,
+                             DatasetIdGenEnum,
+                             DatasetRef,
+                             DatasetId,
+                             ButlerURI,
+                             )
+from lsst.daf.butler.core.utils import getClassOf
+from lsst.daf.butler.transfers import RepoExportContext
+from lsst.daf.butler.registry import CollectionType
+from lsst.daf.butler.datastores.fileDatastore import FileDatastore
+from lsst.daf.butler.registry.databases.sqlite import SqliteDatabase
+
+log = logging.getLogger(__name__)
+
+
+def rewrite_sqlite_registry(source: str) -> None:
+    """Rewrite a SQLite registry as a new registry.
+
+    * Create a new sqlite registry based on the current configuration.
+    * Export all dimension records, dataset types, collection information
+      and import into new registry.
+    * Transfer datasets without moving them.
+
+    Question is whether we get clever and try to move the old registry
+    out of the way and replace it along with a modified butler.yaml making
+    it look like an inplace rewrite.
+
+    """
+
+    # Create the source butler early so we can ask it questions
+    # without assuming things.
+    source_butler = Butler(source, writeable=False)
+
+    # Check that we are really working with a SQLite database.
+    if not isinstance(source_butler.registry._db, SqliteDatabase):
+        raise RuntimeError("This command can only be used on SQLite registries.")
```

Comment:
The core of it is a "transfer everything from butler1 to butler2" but the check is here because I am creating my own registry to copy into and assuming I'm copying it back again.  I'll move the generic transfer code to a separate function at least.